libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -384,27 +374,135 @@ export class CheckoutService {
         }),
       ]);
 
-      const updatedVersion = version + 1;
-
-      if (!subscription.latest_invoice) {
-        throw new CheckoutError(
-          'latest_invoice does not exist on subscription'
-        );
-      }
-      const latestInvoice = await this.invoiceManager.retrieve(
-        subscription.latest_invoice
-      );
-      try {
-        this.invoiceManager.processPayPalInvoice(latestInvoice);
-      } catch (e) {
-        await this.subscriptionManager.cancel(subscription.id);
-        await this.paypalBillingAgreementManager.cancel(billingAgreementId);
-      }
-
-      await this.postPaySteps(cart, updatedVersion, subscription, uid);
+      this.pollAndUpdatePaymentState(cart.id);
     } catch (error) {
       console.error(error);
       throw error;
     }
   }
+
+  async pollAndUpdatePaymentState(cartId: string) {
```

Comment:
Just as a general note for knowledge sharing, with SP3.0 we've made attempts to not have 1 method with all the logic and a bunch of conditional statements, and instead have multiple smaller methods each performing specific functions.

Specifically how this applies to the `CheckoutService`, is that there is a `payWithPaypal` and `payWithStripe` instead of just a general `pay` method.

Knowing that, I'd expect there to be a `pollForPaypal` and `pollForStripe` method, instead of having 1 method for both.

(Note, however, as mentioned in another comment further down, I don't believe PayPal needs a polling method, I just thought this might be a good place to share some of SP3s thinking)